### ISSUE TITLE
User can authenticate with SAML SSO

### ISSUE THREAD

**Issue by francisli** [sfbrigade/bats-server#7]
@julianmclain @billhoke 

Per the call this morning, I'm adding this new issue story regarding SAML SSO.

I think we can go ahead and model a new table for users, i.e. what Bill labeled as "BATS User" in his logical data diagram.

This table can contain all the expected metadata for a given user, i.e something like:

user_uuid: uuid (primary key- internal BATS id)
firstname: string
lastname: string
email: string
hashedpassword: string (used only if needed, i.e. no SSO integration)
subjectid: string (an opaque identifier returned by an SSO provider- i.e. an access token, or uuid, etc)
ssodata: jsonb (a generic data blob for holding any additonal info passed to us from an SSO provider during the auth process, i.e. SAML response assertion)

Then we'll need to set up authentication endpoints and support for SAML SSO and do some testing. I'll see if I can find a SAML Identity Provider that is packaged as a Docker image that we can run, to simulate the SSO process in development.

**Comment by francisli** [comment]
Here's a test SAML IDP we could use:

https://hub.docker.com/r/kristophjunge/test-saml-idp/

**Comment by francisli** [comment]
This article covers everything I'm thinking about doing to set us up for SAML SSO- using the above Docker SAML IDP image to run a test IDP, using the passport authentication library with express, with the passport-saml plug-in library for SAML support, etc:

https://medium.com/disney-streaming/setup-a-single-sign-on-saml-test-environment-with-docker-and-nodejs-c53fc1a984c9

**Comment by francisli** [comment]
@billhoke should we maybe also add an attribute like:

issuperuser: bool (not null, default false)

I feel like we are going to need to create some accounts for ourselves as "superusers" to seed the data in the system...?

**Comment by billhoke** [comment]
Hey @francisli I like the idea of that role. Given that there's a spectrum of possibilities from one boolean per role

**Comment by billhoke** [comment]
Here's a draft table design. My tool is a bit out-of-date, and doesn't support JSONB. Maybe we can just hack that in the DDL until I upgrade ... (never my favorite task) 
![image](https://user-images.githubusercontent.com/69280816/90299967-1b4b7700-de4d-11ea-909b-d0dd63f6a14c.png)

**Comment by francisli** [comment]
@billhoke well, we'll always be able to edit the SQL your code exports in our migration file (and we can add comments for any manual changes).

For the email address column, I might switch to use postgres' CITEXT type for easier unique indexing and searching: https://www.postgresql.org/docs/12/citext.html

For the roles, I might suggest both:

issuperuser: bool
roles: role_enum[]

The reason is that "superuser" is really a special case, overrides everything type of role that is convenient to be able to easily set/unset and check for.

To have the type of flexibility you describe, I propose an array of a custom enum type that we would (eventually) define. That way we have the ability for a user to have a composition of roles, and using an enum would help constrain only to valid values...

**Comment by billhoke** [comment]
Data type changes, uuid default and column addition completed. Modified version of diagram and DDL placed in the related thread on the BAT Channel. ;-) I called the super user column superuserindicator just to be consistent with my "class word" traditional column naming convention. I also did not (yet) include a separate table for roles and a many-to-many association table ... guessing/hoping that won't be needed yet ... (?). Happy to get going on that next week if you like ... just gonna take a quick vacation today, Monday and Tuesday (offline).
